token/core/common/wallet.go

```diff
@@ -0,0 +1,225 @@
+	ListHistoryIssuedTokens() (*token.IssuedTokens, error)
+}
+
+type WalletRegistry interface {
+	WalletIDs() ([]string, error)
+	RegisterIdentity(id string, path string) error
+	Lookup(id interface{}) (driver.Wallet, driver.IdentityInfo, string, error)
+	RegisterWallet(id string, wallet driver.Wallet) error
+	BindIdentity(identity view.Identity, eID string, wID string, meta any) error
+	ContainsIdentity(i view.Identity, id string) bool
+}
+
+type WalletFactory interface {
+	NewWallet(role driver.IdentityRole, walletRegistry WalletRegistry, info driver.IdentityInfo, id string) (driver.Wallet, error)
+}
+
+type DeserializerProviderFunc = func() (driver.Deserializer, error)
+
+type RegistryEntry struct {
+	Registry WalletRegistry
+	Mutex    *sync.RWMutex
+}
+
+type WalletService struct {
+	Logger               *flogging.FabricLogger
+	IdentityProvider     driver.IdentityProvider
+	DeserializerProvider DeserializerProviderFunc
+
+	WalletFactory WalletFactory
+	Registries    map[driver.IdentityRole]*RegistryEntry
+}
+
+func NewWalletService(
+	logger *flogging.FabricLogger,
+	identityProvider driver.IdentityProvider,
+	deserializerProvider DeserializerProviderFunc,
+	walletFactory WalletFactory,
+	OwnerWalletRegistry WalletRegistry,
+	IssuerWalletRegistry WalletRegistry,
+	AuditorWalletRegistry WalletRegistry,
+	CertifierWalletsRegistry WalletRegistry,
+) *WalletService {
+	registries := map[driver.IdentityRole]*RegistryEntry{}
+	registries[driver.OwnerRole] = &RegistryEntry{Registry: OwnerWalletRegistry, Mutex: &sync.RWMutex{}}
+	registries[driver.IssuerRole] = &RegistryEntry{Registry: IssuerWalletRegistry, Mutex: &sync.RWMutex{}}
+	registries[driver.AuditorRole] = &RegistryEntry{Registry: AuditorWalletRegistry, Mutex: &sync.RWMutex{}}
+	registries[driver.CertifierRole] = &RegistryEntry{Registry: CertifierWalletsRegistry, Mutex: &sync.RWMutex{}}
+
+	return &WalletService{
+		Logger:               logger,
+		IdentityProvider:     identityProvider,
+		DeserializerProvider: deserializerProvider,
+		WalletFactory:        walletFactory,
+		Registries:           registries,
+	}
+}
+
+func (s *WalletService) RegisterOwnerWallet(id string, path string) error {
+	return s.Registries[driver.OwnerRole].Registry.RegisterIdentity(id, path)
+}
+
+func (s *WalletService) RegisterIssuerWallet(id string, path string) error {
+	return s.Registries[driver.IssuerRole].Registry.RegisterIdentity(id, path)
+}
+
+func (s *WalletService) GetAuditInfo(id view.Identity) ([]byte, error) {
+	return s.IdentityProvider.GetAuditInfo(id)
+}
+
+func (s *WalletService) GetEnrollmentID(auditInfo []byte) (string, error) {
+	return s.IdentityProvider.GetEnrollmentID(auditInfo)
+}
+
+func (s *WalletService) GetRevocationHandler(auditInfo []byte) (string, error) {
+	return s.IdentityProvider.GetRevocationHandler(auditInfo)
+}
+
+func (s *WalletService) RegisterRecipientIdentity(data *driver.RecipientData) error {
+	if data == nil {
+		return errors.Errorf("cannot recigest empty recipient")
```

Comment:
Not for performance (it's a negligible gain), but maybe as aroule we could define such errors as "constants"

Reply:
the issue here is that `errors.Errorf` records also the stack trace. I can create an error and then I will have to wrap it there to add the stack trace. Would that be okay?

Reply:
aaah ok fair!